examples/03_evaluate/als_movielens_diversity_metrics.ipynb

```diff
@@ -722,85 +732,239 @@
+      "+-------+---------------------------------------------+\n",
+      "only showing top 10 rows\n",
+      "\n"
+     ]
+    }
+   ],
+   "metadata": {}
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "The *features* column is represented with a SparseVector object. For example, in the feature vector (1043,[128,544,1025],[1.0,1.0,1.0]), 1043 is the vector length, indicating the vector consisting of 1043 item features. The values at index positions 128,544,1025 are 1.0, and the values at other positions are all 0. "
+   ],
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 30,
+   "source": [
+    "als_diversity_eval = SparkDiversityEvaluation(\n",
+    "    train_df = train_df, \n",
+    "    reco_df = top_k_reco,\n",
+    "    item_feature_df = feature_data, \n",
+    "    col_user = COL_USER, \n",
+    "    col_item = COL_ITEM\n",
+    ")\n",
+    "\n",
+    "als_diversity=als_diversity_eval.diversity()\n",
+    "als_diversity"
+   ],
+   "outputs": [
+    {
+     "output_type": "execute_result",
+     "data": {
       "text/plain": [
-       "   Data    Algo   K  Precision@k  Recall@k    NDCG@k  Mean average precision  \\\n",
-       "1  100k     als  10     0.051911  0.017514  0.047460                0.005734   \n",
-       "2  100k  random  10     0.015393  0.005420  0.016805                0.002053   \n",
-       "\n",
-       "   catalog_coverage  distributional_coverage    novelty  diversity  \\\n",
-       "1          0.381906                 8.029716  11.800048   0.892449   \n",
-       "2          0.998179                10.557924  12.179461   0.923035   \n",
-       "\n",
-       "   serendipity  \n",
-       "1     0.880236  \n",
-       "2     0.893947  "
+       "0.8738984131037538"
       ]
      },
-     "execution_count": 21,
      "metadata": {},
-     "output_type": "execute_result"
+     "execution_count": 30
     }
    ],
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 31,
    "source": [
-    "df_results"
-   ]
+    "random_diversity_eval = SparkDiversityEvaluation(\n",
+    "    train_df = train_df, \n",
+    "    reco_df = pred_df, \n",
+    "    item_feature_df = feature_data, \n",
+    "    col_user = COL_USER, \n",
+    "    col_item = COL_ITEM\n",
+    ")\n",
+    "  \n",
+    "random_diversity =random_diversity_eval.diversity()\n",
+    "random_diversity"
+   ],
+   "outputs": [
+    {
+     "output_type": "execute_result",
+     "data": {
+      "text/plain": [
+       "0.8970750054249245"
+      ]
+     },
+     "metadata": {},
+     "execution_count": 31
+    }
+   ],
+   "metadata": {}
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "It's interesting that diversity changes when using different item-item similarity calculation approach, for both ALS algorithm and random recommender. But the diversity of random recommender is still higher than ALS algorithm. "
+   ],
+   "metadata": {}
   },
```

Comment:
Another thing you could add here is how serendipity changes now that cosine similarity is computed using the item feature vectors.

Reply:
added calculation for serendipity.